File: #10 Matrix/matrix.py
def print_matrix(matrix):
    for row in matrix:
        print(*row)
    print()

def sort_row_wise(matrix):
    for row in matrix:
        row.sort()

def sort_column_wise(matrix):
    rows = len(matrix)
    cols = len(matrix[0])

    for col in range(cols):
        temp = []

        for row in range(rows):
            temp.append(matrix[row][col])

        temp.sort()

        for row in range(rows):
            matrix[row][col] = temp[row]

def rotate_clockwise_by_one(matrix):
    rows = len(matrix)
    cols = len(matrix[0])

    top = 0
    bottom = rows - 1
    left = 0
    right = cols - 1

    prev = matrix[top + 1][left]

    for i in range(left, right + 1):
        matrix[top][i], prev = prev, matrix[top][i]
    top += 1

    for i in range(top, bottom + 1):
        matrix[i][right], prev = prev, matrix[i][right]
    right -= 1

    for i in range(right, left - 1, -1):
        matrix[bottom][i], prev = prev, matrix[bottom][i]
    bottom -= 1

    for i in range(bottom, top - 1, -1):
        matrix[i][left], prev = prev, matrix[i][left]

def rotate_counter_clockwise_by_one(matrix):
    rows = len(matrix)
    cols = len(matrix[0])

    top = 0
    bottom = rows - 1
    left = 0
    right = cols - 1

    prev = matrix[top][left + 1]

    for i in range(top, bottom + 1):
        matrix[i][left], prev = prev, matrix[i][left]
    left += 1

    for i in range(left, right + 1):
        matrix[bottom][i], prev = prev, matrix[bottom][i]
    bottom -= 1

    for i in range(bottom, top - 1, -1):
        matrix[i][right], prev = prev, matrix[i][right]
    right -= 1

    for i in range(right, left - 1, -1):
        matrix[top][i], prev = prev, matrix[top][i]

def rotate_90(matrix):
    return [list(row) for row in zip(*matrix[::-1])]

def rotate_180(matrix):
    return [row[::-1] for row in matrix[::-1]]

def row_wise_traversal(matrix):
    for row in matrix:
        for val in row:
            print(val, end=" ")
    print()

def column_wise_traversal(matrix):
    rows = len(matrix)
    cols = len(matrix[0])

    for col in range(cols):
        for row in range(rows):
            print(matrix[row][col], end=" ")
    print()

def spiral_print(matrix):
    top = 0
    bottom = len(matrix) - 1
    left = 0
    right = len(matrix[0]) - 1

    while top <= bottom and left <= right:

        for i in range(left, right + 1):
            print(matrix[top][i], end=" ")
        top += 1

        for i in range(top, bottom + 1):
            print(matrix[i][right], end=" ")
        right -= 1

        if top <= bottom:
            for i in range(right, left - 1, -1):
                print(matrix[bottom][i], end=" ")
            bottom -= 1

        if left <= right:
            for i in range(bottom, top - 1, -1):
                print(matrix[i][left], end=" ")
            left += 1

    print()

def transpose(matrix):
    return [list(row) for row in zip(*matrix)]

rows = int(input("Jumlah baris: "))
cols = int(input("Jumlah kolom: "))

matrix = []

print("Input matrix:")
for i in range(rows):
    row = list(map(int, input().split()))
    matrix.append(row)

while True:
    print("\nMENU")
    print("1-a. Sort row-wise")
    print("1-b. Sort column-wise")
    print("2-a. Rotate clockwise by 1")
    print("2-b. Rotate counter-clockwise by 1")
    print("2-c. Rotate 90")
    print("2-d. Rotate 180")
    print("3-a. Row-wise traversal")
    print("3-b. Column-wise traversal")
    print("4. Spiral form")
    print("5. Transpose")
    print("6. Quit")

    choice = input("Pilih menu: ")

    if choice == "1a":
        sort_row_wise(matrix)
        print_matrix(matrix)

    elif choice == "1b":
        sort_column_wise(matrix)
        print_matrix(matrix)

    elif choice == "2a":
        rotate_clockwise_by_one(matrix)
        print_matrix(matrix)

    elif choice == "2b":
        rotate_counter_clockwise_by_one(matrix)
        print_matrix(matrix)

    elif choice == "2c":
        matrix = rotate_90(matrix)
        print_matrix(matrix)

    elif choice == "2d":
        matrix = rotate_180(matrix)
        print_matrix(matrix)

    elif choice == "3a":
        row_wise_traversal(matrix)

    elif choice == "3b":
        column_wise_traversal(matrix)

    elif choice == "4":
        spiral_print(matrix)

    elif choice == "5":
        matrix = transpose(matrix)
        print_matrix(matrix)

    elif choice == "6":
        print("Program selesai")
        break

    else:
        print("Pilihan tidak valid")

File: #10 Matrix/matrix.java
import java.util.*;

public class matrix {

    static Scanner sc = new Scanner(System.in);

    static void printMatrix(int[][] matrix) {
        for (int[] row : matrix) {
            for (int val : row) {
                System.out.print(val + " ");
            }
            System.out.println();
        }
        System.out.println();
    }

    static void sortRowWise(int[][] matrix) {
        for (int[] row : matrix) {
            Arrays.sort(row);
        }
    }

    static void sortColumnWise(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        for (int col = 0; col < cols; col++) {

            int[] temp = new int[rows];

            for (int row = 0; row < rows; row++) {
                temp[row] = matrix[row][col];
            }

            Arrays.sort(temp);

            for (int row = 0; row < rows; row++) {
                matrix[row][col] = temp[row];
            }
        }
    }

    static int[][] rotate90(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        int[][] result = new int[cols][rows];

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                result[j][rows - 1 - i] = matrix[i][j];
            }
        }

        return result;
    }

    static int[][] rotate180(int[][] matrix) {
        return rotate90(rotate90(matrix));
    }

    static void rowTraversal(int[][] matrix) {
        for (int[] row : matrix) {
            for (int val : row) {
                System.out.print(val + " ");
            }
        }
        System.out.println();
    }

    static void columnTraversal(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;

        for (int j = 0; j < cols; j++) {
            for (int i = 0; i < rows; i++) {
                System.out.print(matrix[i][j] + " ");
            }
        }
        System.out.println();
    }

    static void spiralPrint(int[][] matrix) {

        int top = 0;
        int bottom = matrix.length - 1;
        int left = 0;
        int right = matrix[0].length - 1;

        while (top <= bottom && left <= right) {

            for (int i = left; i <= right; i++)
                System.out.print(matrix[top][i] + " ");
            top++;

            for (int i = top; i <= bottom; i++)
                System.out.print(matrix[i][right] + " ");
            right--;

            if (top <= bottom) {
                for (int i = right; i >= left; i--)
                    System.out.print(matrix[bottom][i] + " ");
                bottom--;
            }

            if (left <= right) {
                for (int i = bottom; i >= top; i--)
                    System.out.print(matrix[i][left] + " ");
                left++;
            }
        }

        System.out.println();
    }

    static int[][] transpose(int[][] matrix) {

        int rows = matrix.length;
        int cols = matrix[0].length;

        int[][] result = new int[cols][rows];

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                result[j][i] = matrix[i][j];
            }
        }

        return result;
    }

    public static void main(String[] args) {

        System.out.print("Jumlah baris: ");
        int rows = sc.nextInt();

        System.out.print("Jumlah kolom: ");
        int cols = sc.nextInt();

        int[][] matrix = new int[rows][cols];

        System.out.println("Input matrix:");

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                matrix[i][j] = sc.nextInt();
            }
        }

        while (true) {

            System.out.println("\nMENU");
            System.out.println("1a. Sort row-wise");
            System.out.println("1b. Sort column-wise");
            System.out.println("2c. Rotate 90");
            System.out.println("2d. Rotate 180");
            System.out.println("3a. Row-wise traversal");
            System.out.println("3b. Column-wise traversal");
            System.out.println("4. Spiral form");
            System.out.println("5. Transpose");
            System.out.println("6. Quit");

            System.out.print("Pilih menu: ");
            String choice = sc.next();

            switch (choice) {

                case "1a":
                    sortRowWise(matrix);
                    printMatrix(matrix);
                    break;

                case "1b":
                    sortColumnWise(matrix);
                    printMatrix(matrix);
                    break;

                case "2c":
                    matrix = rotate90(matrix);
                    printMatrix(matrix);
                    break;

                case "2d":
                    matrix = rotate180(matrix);
                    printMatrix(matrix);
                    break;

                case "3a":
                    rowTraversal(matrix);
                    break;

                case "3b":
                    columnTraversal(matrix);
                    break;

                case "4":
                    spiralPrint(matrix);
                    break;

                case "5":
                    matrix = transpose(matrix);
                    printMatrix(matrix);
                    break;

                case "6":
                    System.out.println("Program selesai");
                    return;

                default:
                    System.out.println("Pilihan tidak valid");
            }
        }
    }
}